File: Editorial/95_Unique_Binary_Search_Trees_II/JavaScript/1_Recursive_Dynamic_Programming.js
var allPossibleBST = function (start, end, memo) {
    let res = [];
    if (start > end) {
        res.push(null);
        return res;
    }
    let key = start + "," + end;
    if (memo[key] != undefined) {
        return memo[key];
    }
    for (let i = start; i <= end; ++i) {
        let leftSubTrees = allPossibleBST(start, i - 1, memo);
        let rightSubTrees = allPossibleBST(i + 1, end, memo);
        for (let j = 0; j < leftSubTrees.length; j++) {
            for (let k = 0; k < rightSubTrees.length; k++) {
                let root = new TreeNode(i, leftSubTrees[j], rightSubTrees[k]);
                res.push(root);
            }
        }
    }
    memo[key] = res;
    return res;
};
var generateTrees = function (n) {
    let memo = {};
    return allPossibleBST(1, n, memo);
};

File: Editorial/0000-0100/0095_Unique_Binary_Search_Trees_II/Go/1_Recursive_Dynamic_Programming.go
func allPossibleBST(
    start int,
    end int,
    memo map[string][]*TreeNode,
) []*TreeNode {
    res := []*TreeNode{}
    if start > end {
        res = append(res, nil)
        return res
    }
    key := fmt.Sprintf("%d,%d", start, end)
    if value, ok := memo[key]; ok {
        return value
    }
    // Iterate through all values from start to end to construct left and right subtree recursively.
    for i := start; i <= end; i++ {
        leftSubTrees := allPossibleBST(start, i-1, memo)
        rightSubTrees := allPossibleBST(i+1, end, memo)
        // Loop through all left and right subtrees and connect them to ith root.
        for _, left := range leftSubTrees {
            for _, right := range rightSubTrees {
                root := TreeNode{Val: i, Left: left, Right: right}
                res = append(res, &root)
            }
        }
    }
    memo[key] = res
    return res
}

func generateTrees(n int) []*TreeNode {
    memo := make(map[string][]*TreeNode)
    return allPossibleBST(1, n, memo)
}
